File: app/src/main/java/dev/ragnarok/fenrir/api/model/AttachmentsTokenCreator.java
package dev.ragnarok.fenrir.api.model;


public class AttachmentsTokenCreator {

    public static IAttachmentToken ofDocument(int id, int ownerId, String accessKey) {
        return new AttachmentToken("doc", id, ownerId, accessKey);
    }

    public static IAttachmentToken ofAudio(int id, int ownerId, String accessKey) {
        return new AttachmentToken("audio", id, ownerId, accessKey);
    }

    public static IAttachmentToken ofLink(String url) {
        return new LinkAttachmentToken(url);
    }

    public static IAttachmentToken ofArticle(int id, int ownerId, String accessKey) {
        return new AttachmentToken("article", id, ownerId, accessKey);
    }

    public static IAttachmentToken ofStory(int id, int ownerId, String accessKey) {
        return new AttachmentToken("story", id, ownerId, accessKey);
    }

    public static IAttachmentToken ofPhotoAlbum(int id, int ownerId) {
        return new AttachmentToken("album", id, ownerId);
    }

    public static IAttachmentToken ofAudioPlaylist(int id, int ownerId, String accessKey) {
        return new AttachmentToken("audio_playlist", id, ownerId, accessKey);
    }

    public static IAttachmentToken ofGraffity(int id, int ownerId, String accessKey) {
        return new AttachmentToken("graffiti", id, ownerId, accessKey);
    }

    public static IAttachmentToken ofCall(int initiator_id, int receiver_id, String state, long time) {
        return new AttachmentToken("call", initiator_id, receiver_id, state + "_" + time);
    }

    public static IAttachmentToken ofPhoto(int id, int ownerId, String accessKey) {
        return new AttachmentToken("photo", id, ownerId, accessKey);
    }

    public static IAttachmentToken ofPoll(int id, int ownerId) {
        return new AttachmentToken("poll", id, ownerId);
    }

    public static IAttachmentToken ofWallReply(int id, int ownerId) {
        return new AttachmentToken("wall_reply", id, ownerId);
    }

    public static IAttachmentToken ofPost(int id, int ownerId) {
        return new AttachmentToken("wall", id, ownerId);
    }

    public static IAttachmentToken ofError(String type, String data) {
        return new AttachmentToken("error", type.hashCode(), data.hashCode());
    }

    public static IAttachmentToken ofEvent(int id, String button_text) {
        return new AttachmentToken("event", id, button_text.hashCode());
    }

    public static IAttachmentToken ofMarket(int id, int ownerId, String accessKey) {
        return new AttachmentToken("market", id, ownerId, accessKey);
    }

    public static IAttachmentToken ofMarketAlbum(int id, int ownerId, String accessKey) {
        return new AttachmentToken("market_album", id, ownerId, accessKey);
    }

    public static IAttachmentToken ofVideo(int id, int ownerId, String accessKey) {
        return new AttachmentToken("video", id, ownerId, accessKey);
    }

    public static IAttachmentToken ofArtist(String id) {
        return new AttachmentTokenString("artist", id);
    }
}

File: app_fenrir/src/main/kotlin/dev/ragnarok/fenrir/api/model/AttachmentsTokenCreator.kt
package dev.ragnarok.fenrir.api.model

import dev.ragnarok.fenrir.api.model.interfaces.IAttachmentToken

object AttachmentsTokenCreator {
    fun ofDocument(id: Int, ownerId: Long, accessKey: String?): IAttachmentToken {
        return AttachmentTokens.AttachmentToken("doc", id, ownerId, accessKey)
    }

    fun ofAudio(id: Int, ownerId: Long, accessKey: String?): IAttachmentToken {
        return AttachmentTokens.AttachmentToken("audio", id, ownerId, accessKey)
    }

    fun ofLink(url: String?): IAttachmentToken {
        return AttachmentTokens.LinkAttachmentToken(url)
    }

    fun ofArticle(id: Int, ownerId: Long, accessKey: String?): IAttachmentToken {
        return AttachmentTokens.AttachmentToken("article", id, ownerId, accessKey)
    }

    fun ofStory(id: Int, ownerId: Long, accessKey: String?): IAttachmentToken {
        return AttachmentTokens.AttachmentToken("story", id, ownerId, accessKey)
    }

    fun ofPhotoAlbum(id: Int, ownerId: Long): IAttachmentToken {
        return AttachmentTokens.AttachmentToken("album", id, ownerId)
    }

    fun ofAudioPlaylist(id: Int, ownerId: Long, accessKey: String?): IAttachmentToken {
        return AttachmentTokens.AttachmentToken("audio_playlist", id, ownerId, accessKey)
    }

    fun ofGraffiti(id: Int, ownerId: Long, accessKey: String?): IAttachmentToken {
        return AttachmentTokens.AttachmentToken("graffiti", id, ownerId, accessKey)
    }

    fun ofCall(
        initiator_id: Long,
        receiver_id: Long,
        state: String?,
        time: Long
    ): IAttachmentToken {
        return AttachmentTokens.AttachmentTokenStringSpecial(
            "call",
            initiator_id.toString(),
            receiver_id.toString() + "_{$state}_$time"
        )
    }

    fun ofGeo(latitude: String?, longitude: String?): IAttachmentToken {
        return AttachmentTokens.AttachmentToken(
            "geo",
            latitude.hashCode(),
            longitude.hashCode().toLong()
        )
    }

    fun ofPhoto(id: Int, ownerId: Long, accessKey: String?): IAttachmentToken {
        return AttachmentTokens.AttachmentToken("photo", id, ownerId, accessKey)
    }

    fun ofPoll(id: Int, ownerId: Long): IAttachmentToken {
        return AttachmentTokens.AttachmentToken("poll", id, ownerId)
    }

    fun ofWallReply(id: Int, ownerId: Long): IAttachmentToken {
        return AttachmentTokens.AttachmentToken("wall_reply", id, ownerId)
    }

    fun ofPost(id: Int, ownerId: Long): IAttachmentToken {
        return AttachmentTokens.AttachmentToken("wall", id, ownerId)
    }

    fun ofError(type: String?, data: String?): IAttachmentToken {
        return AttachmentTokens.AttachmentToken("error", type.hashCode(), data.hashCode().toLong())
    }

    fun ofEvent(id: Long, button_text: String?): IAttachmentToken {
        return AttachmentTokens.AttachmentToken(
            "event",
            id.hashCode(),
            button_text.hashCode().toLong()
        )
    }

    fun ofMarket(id: Int, ownerId: Long, accessKey: String?): IAttachmentToken {
        return AttachmentTokens.AttachmentToken("market", id, ownerId, accessKey)
    }

    fun ofMarketAlbum(id: Int, ownerId: Long, accessKey: String?): IAttachmentToken {
        return AttachmentTokens.AttachmentToken("market_album", id, ownerId, accessKey)
    }

    fun ofVideo(id: Int, ownerId: Long, accessKey: String?): IAttachmentToken {
        return AttachmentTokens.AttachmentToken("video", id, ownerId, accessKey)
    }

    fun ofArtist(id: String?): IAttachmentToken {
        return AttachmentTokens.AttachmentTokenForArtist("artist", id)
    }
}